People/app/src/main/java/com/example/android/people/ui/main/MainFragment.kt

```diff
@@ -45,17 +52,35 @@ class MainFragment : Fragment(R.layout.main_fragment) {
         navigationController.updateAppBar(false)
         val viewModel: MainViewModel by viewModels()
 
+        // Show a header message asking the user to grant the notification permission.
+        val headerAdapter = HeaderAdapter { permissionRequest.launch() }
+
+        // Show the contact list.
         val contactAdapter = ContactAdapter { id ->
             navigationController.openChat(id, null)
         }
-        viewModel.contacts.observe(viewLifecycleOwner, Observer { contacts ->
-            contactAdapter.submitList(contacts)
-        })
         binding.contacts.run {
             layoutManager = LinearLayoutManager(view.context)
             setHasFixedSize(true)
             adapter = contactAdapter
         }
+        viewModel.contacts.observe(viewLifecycleOwner) { contacts ->
+            contactAdapter.submitList(contacts)
+        }
+
+        // Deal with updates of the permission status.
+        permissionRequest.status.observe(viewLifecycleOwner) { status ->
+            when (status) {
+                // We have the permission now. Hide the permission header.
+                is PermissionStatus.Granted -> binding.contacts.adapter = contactAdapter
+                // We don't have the permission. Show the permission header.
+                is PermissionStatus.Denied -> {
+                    binding.contacts.adapter = ConcatAdapter(headerAdapter, contactAdapter)
```

Comment:
This makes `contacts.adapter` has no stable ids

Reply:
That's a very good point. I'll fix it.

Reply:
#413 🙂

Reply:
Merged. Thank you so much for fix it.